### PR TITLE
KOGITO-6289: Regressions on copy/paste from other spreadsheet, after KOGITO-5078

### PR DESCRIPTION
**Pull request by ljmotta**
## JIRA
https://issues.redhat.com/browse/KOGITO-6289

I've fixed the issue described in the JIRA, and also, I've removed most of the listeners added to the document, which should be reduced to the `boxed-expression-component` scope. This was done to avoid undesired listeners where the Boxed Expression Component is rendered.

### CODE REVIEW

File: kogito-editors-js/packages/boxed-expression-component/src/components/SelectionBox/SelectionBox.tsx
Comment (tiagobento) [inline kogito-editors-js/packages/boxed-expression-component/src/components/SelectionBox/SelectionBox.tsx:70]: Great job removing a `useEffect`. The less of them, the better!

File: kogito-editors-js/packages/boxed-expression-component/src/components/Table/Table.tsx
Comment (tiagobento) [inline kogito-editors-js/packages/boxed-expression-component/src/components/Table/Table.tsx:189]: I feel like this function should have an object parameter. It's unclear to me what those two sequential `undefined`s mean. Can you please change the signature of this method to receive a single object-parameter instead of 4 separate parameters?

File: kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx
Comment (tiagobento) [inline kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx:79]: Great catch. We should avoid using global listeners whenever possible. I understand that this is out of scope for this PR, but kudos for boy-scouting. Could you please create a Jira for this issue and link the Jira number on this comment.

File: kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx
Comment (tiagobento) [inline kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx:47]: I think we should avoid using `ids` for the case where we need more than one instance of BoxedExpressionComponent in a page. Can you please change this to `className`?

File: kogito-editors-js/packages/boxed-expression-component/src/api/BoxedExpressionEditor.ts
Comment (tiagobento) [inline kogito-editors-js/packages/boxed-expression-component/src/api/BoxedExpressionEditor.ts:85]: See my comment about `id`s.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Thanks for the PR, Luiz! Left some comments inline.

File: kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx
Comment (tiagobento) [reply kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx:47]: Ok, I just realized using `className` won't work too :/ The solution would be having the consumer of BoxedExpressionComponent set this `id`, delegating the responsibility of maintaining the uniqueness of this `id` to the consumer, or use a `ref` for this `div` and pass it along using props or even a context. WDYT?

File: kogito-editors-js/packages/boxed-expression-component/src/components/Table/Table.tsx
Comment (ljmotta) [reply kogito-editors-js/packages/boxed-expression-component/src/components/Table/Table.tsx:189]: Yes, sure!

File: kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx
Comment (ljmotta) [reply kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx:79]: Ok! I'm going to create a JIRA to track this.

File: kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx
Comment (ljmotta) [reply kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.tsx:47]: It would be a good solution. I'm going to use the` BoxedExpressionGlobalContext` to pass it down on the tree, and remove the `id`.

**Review by yesamer — COMMENTED** [review]
@ljmotta It seems there's a conflict here, can you please resolve it? Thx

File: kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx
Comment (ljmotta) [reply kogito-editors-js/packages/boxed-expression-component/src/components/Resizer/Resizer.tsx]: Actually, I've manage to fix the issue already.

**Comment by ljmotta** [comment]
@tiagobento I've done all your change requests, could you review it again?

@yesamer and conflicts solved.

**Review by tiagobento — APPROVED** [review]
Thanks for applying the suggested changes, Luiz!

**Comment by ljmotta** [comment]
@yesamer @vpellegrino @jomarko Hi, I believe I've done the necessary adjustments, this PR is ready for review.

**Review by vpellegrino — COMMENTED** [review]
@ljmotta nice work!
From a code perspective, it looks good to me.
I just need a few more hours to test all touched use-cases and the integration with the DMN Editor.

**Comment by yesamer** [comment]
Jenkins retest this

**Comment by vpellegrino** [comment]
@ljmotta, it seems that copy&paste is working fine (or at least as before).
I just noticed a different user experience when selecting multiple cells.
When selecting multiple cells, with an older version of the editor (https://boxed-expression-editor.netlify.app/) the selection seems more fluid, even when the mouse cursor goes out of the area of the cells:
fluid_selection

Instead, with the editor produced by this PR, it seems we have a glitch when the mouse cursor goes out of the area of the table cells. I tried to capture it with a GIF, but I am not sure that it helps to understand:
selection_box_glitch.
In such situations, no cell is selected, and the selection box is still active until you click outside somewhere. Moreover, it seems that the mouse cursor is a bit slow while performing the selection action.

This selection issue does not seem critical, but I wanted to let you know.
In the meantime, please let me know if you need further details.

**Review by jomarko — COMMENTED** [review]
@vpellegrino @ljmotta 

Hello, some good news here. I was able to automate a test for copy and paste, see https://github.com/ljmotta/kogito-editors-java/pull/2.

Right now I did it just for a relation.

@vpellegrino could you please provide a list of expressions, where such copy and paste should be possible? I guess supported it is in:
- Decision Table
- Relation
- List

Is that correct?

Once I know, I will extend the coverage.

May I ask you to review and possible merge my copy and paste test? If steps to run locally needed here it is:
- $ kogito-editors-js - yarn // to install dependencies
- $ kogito-editors-js - yarn build:dev
- $ boxed-expression-editor/showcase - yarn start
- $ boxed-expression-editor/showcase - yarn cypress open - and pick any test you want to run

**Comment by vpellegrino** [comment]
Answering to https://github.com/kiegroup/kogito-editors-java/pull/181#pullrequestreview-827862534
@jomarko in the original task https://issues.redhat.com/browse/KOGITO-3656 there is no specific indication of supported expressions.
I tried a bit https://boxed-expression-editor.netlify.app/ and it seems that:
- Literal
- Relation
- Decision Table
are supporting the copy/paste feature.
The same applies also to nested expressions that wrap a text cell (for instance, a context table containing a literal expression for one of its entries).

I think that this PR (https://github.com/kiegroup/kogito-editors-java/pull/103) clarifies better possible use-cases.

Tomorrow I'll review your test cases PR, thanks a lot @jomarko

**Comment by ljmotta** [comment]
@vpellegrino Thanks for your review. Indeed, now the selection box can only exist inside the `boxed-expression-provider` `div` instead of the `document` itself. So to increase the selection box range, it's necessary to have a bigger parent `div`. I believe that I've managed to increase the DMN Boxed Editor on the DMN Editor, but it would be good if you could check if I have changed the CSS in the correct place. Here is a quick video using the `kie-wb-common-dmn-webapp-kogito-testing`:

https://user-images.githubusercontent.com/24302289/145454470-ff315b72-acfa-48fc-99cf-1251c2cf7719.mov

@jomarko Thanks a lot for your test case, I'm going to review it.

**Comment by yesamer** [comment]
Jenkins retest this

**Review by vpellegrino — APPROVED** [review]
@ljmotta, great!
Thanks for finding a solution for it.
Do you think it is possible to fix the `showcase` app as well? In that case the HTML structure it's a bit different.

**Comment by ljmotta** [comment]
@jomarko I tried the tests locally and they look great. Thanks!

@vpellegrino What is possible to do is to increase the Boxed Editor `div` height, so it will have a bigger gap between the response and the Editor itself. This will not enable to select on top of the response but will increase the selection area. For example:

image

**Comment by ljmotta** [comment]
@vpellegrino I've set the height of the showcase to make the selection area bigger. Adding to that, I've added a dashed border to make it possible to see it.

**Comment by yesamer** [comment]
jenkins retest this please

**Comment by yesamer** [comment]
@ljmotta  jenkins build reported this failure: WDYT?

[2021-12-13T13:03:12.668Z] [INFO] boxed-expression-component: $ yarn test && webpack --config webpack.production.js
[2021-12-13T13:03:12.668Z] [INFO] boxed-expression-component-showcase: $ yarn test:it
[2021-12-13T13:03:12.668Z] [INFO] boxed-expression-component: $ jest --collect-coverage
[2021-12-13T13:03:12.668Z] [INFO] boxed-expression-component-showcase: $ start-server-and-test start:it http://localhost:3000 cy:run
[2021-12-13T13:03:22.848Z] [INFO] boxed-expression-component-showcase: 1: starting server using command "npm run start:it"
[2021-12-13T13:03:22.848Z] [INFO] boxed-expression-component-showcase: and when url "[ 'http://localhost:3000' ]" is responding with HTTP status code 200
[2021-12-13T13:03:22.848Z] [INFO] boxed-expression-component-showcase: running tests using command "npm run cy:run"
[2021-12-13T13:03:23.139Z] [INFO] boxed-expression-component-showcase: > boxed-expression-component-showcase@0.0.0 start:it
[2021-12-13T13:03:23.139Z] [INFO] boxed-expression-component-showcase: > BROWSER=none yarn start
[2021-12-13T13:03:23.439Z] [INFO] boxed-expression-component-showcase: $ react-app-rewired start
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: ℹ ｢wds｣: Project is running at http://10.0.100.195/
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: ℹ ｢wds｣: webpack output is served from 
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: ℹ ｢wds｣: Content not from webpack is served from /home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/packages/boxed-expression-component/showcase/public
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: ℹ ｢wds｣: 404s will fallback to /
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: Starting the development server...
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: node:events:342
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase:       throw er; // Unhandled 'error' event
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase:       ^
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase: Error: ENOSPC: System limit for number of file watchers reached, watch '/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/packages/boxed-expression-component/showcase/public/favicon.ico'
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase:     at FSWatcher.<computed> (node:internal/fs/watchers:244:19)
[2021-12-13T13:03:45.800Z] [INFO] boxed-expression-component-showcase:     at Object.watch (node:fs:2240:34)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at createFsWatchInstance (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:38:15)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at setFsWatchListener (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:233:14)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSWatcher.NodeFsHandler._handleFile (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:262:21)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSWatcher.<anonymous> (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:495:21)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSReqCallback.oncomplete (node:fs:195:5)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: Emitted 'error' event on FSWatcher instance at:
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSWatcher._handleError (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/index.js:260:10)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at createFsWatchInstance (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:40:5)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at setFsWatchListener (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     [... lines matching original stack trace ...]
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at FSReqCallback.oncomplete (node:fs:195:5) {
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:   errno: -28,
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:   syscall: 'watch',
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:   code: 'ENOSPC',
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:   path: '/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/packages/boxed-expression-component/showcase/public/favicon.ico',
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:   filename: '/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/packages/boxed-expression-component/showcase/public/favicon.ico'
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: }
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: error Command failed with exit code 1.
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: Error: server closed unexpectedly
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at ChildProcess.onClose (/home/jenkins/workspace/KIE/kogito/main/pullrequest/kogito-editors-java.tests.kogito-tooling-editors-java/kogito-editors-java/kogito-editors-js/node_modules/start-server-and-test/src/index.js:75:14)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at ChildProcess.emit (node:events:365:28)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at maybeClose (node:internal/child_process:1067:16)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase:     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: error Command failed with exit code 1.
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: error Command failed with exit code 1.
[2021-12-13T13:03:45.801Z] [INFO] boxed-expression-component-showcase: info Visit https://yarnpkg.com/en/docs/cli/run f

**Comment by jomarko** [comment]
@yesamer oh, thank you for spotting this, I spotted somethig similar during kogito-tooling PRs review.

See:
- https://issues.redhat.com/browse/KOGITO-6355

So I can fix the test. The question is if we need to report new jira or is it the same as https://issues.redhat.com/browse/KOGITO-6355 @ljmotta ?

**Comment by jomarko** [comment]
locally boxed-expression-component pass for me, however not sure if the branch shouldn't be rebased with main so we have most accurate CI results.

**Comment by yesamer** [comment]
jenkins retest this

**Comment by ljmotta** [comment]
@yesamer It looks like this particular test is passing now, but a test failed on tooling. I'm going to take a look into it.

**Review by yesamer — APPROVED** [review]
@ljmotta Thank you.

File: kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.css
Comment (jomarko) [inline kogito-editors-js/packages/boxed-expression-component/src/components/BoxedExpressionEditor/BoxedExpressionProvider.css:17]: I think this change will break some parts of `kogito-tooling`

### Code
- https://github.com/kiegroup/kogito-tooling/blob/main/packages/unitables/src/dmn/style.css#L15

### Tests
- https://github.com/kiegroup/kogito-tooling/blob/main/packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/dmn/ExpressionEditorHelper.ts#L50

**Review by jomarko — COMMENTED** [review]
I am not sure we can merge this without additional changes in `kogito-tooling`, see my comment inline.

**Comment by ljmotta** [comment]
@jomarko Thanks for the heads up. I've opened the [PR](https://github.com/kiegroup/kogito-tooling/pull/718) on tooling to fix this CSS class.

**Comment by yesamer** [comment]
I am afraid we need to release a new version of `kogito-editor-java` to have green lights here. 
@ljmotta Please be sure these two PRs work well in your local, then I'll merge this one even if it fails. After that, we can release a new version with this change included. `kogito-tooling` will refer to this new version and it will be possible to fix all. 
As alternative, we can postpone this merge after the monorepo is in place 
//cc @tiagobento WDTY?

**Comment by tiagobento** [comment]
@yesamer If the CI is green, I'm +1 to merge this, as it will only affect `kogito-tooling` when it is updated with a new version of `kogito-editors-java`. Luiz's PR on `kogito-tooling` should make the CI green here, so no need to wait if everything is green.

**Comment by yesamer** [comment]
Jenkins retest this

**Comment by tiagobento** [comment]
Jenkins retest this

**Review by jomarko — APPROVED** [review]
Thank you for the PR @ljmotta I found these issues that I reported separately to not block this PR from merge:
- [KOGITO-6345](https://issues.redhat.com/browse/KOGITO-6345) - in copied area cell with miltiline text
- [KOGITO-6346](https://issues.redhat.com/browse/KOGITO-6346) copy and paste single cell
- [KOGITO-6347](https://issues.redhat.com/browse/KOGITO-6347) copy and paste in context expression
- [KOGITO-6348](https://issues.redhat.com/browse/KOGITO-6348) empty cells in copied area

Two of them are critical for copy and paste scenario from my point of view.

**Pull request merged**
